### ISSUE TITLE
Some functions could perform type intersection

### ISSUE THREAD

**Issue by Veedrac** [crystal-lang/crystal#2404]
`Set`'s `&`, as an example, is such that

    typeof(Set.new([1, "a"]) & Set.new(["a", false]))

is

    Set(Int32 | String)

which is the type of the LHS, but surely it should really be

    Set(String)

This is possible to implement as

    def intersect(lhs : Set(T), rhs : Set(U))
        set = Set(typeof(nil as Object+ as T as U)).new
        lhs.each do |value|
            if value.is_a?(U) && rhs.includes?(value)
                set.add value
            end
        end
        set
    end

`typeof(nil as Object+ as T as U)` isn't the most obvious syntax (I'd much rather `T & U`), but other than that the code works great.

Would this be useful?

**Comment by asterite** [comment]
Intersection types are a nice idea. Some days ago @bcardiff suggested to me that `-` and `&` should maybe be added to the type grammar. For example `Enumerable#compact` could be defined to return `T - Nil`. And another use case for intersections is this:

```crystal
# We want something that's an IO, but we also want 
# T as a free variable to get the real type
def foo(x : T & IO)
  # do something with T
end
```

Using `Object+` to accomplish that, as commented [here](https://github.com/crystal-lang/crystal/issues/2403#issuecomment-205003215) won't work in the future, but if we add intersection types it could be done.

**Comment by Veedrac** [comment]
Fair enough about `Object+`; I've replaced it with a wordier alternative. Having proper intersection types would be really nice.

**Comment by waterlink** [comment]
I would love to see type intersection too. :+1:

I can only imagine how useful it will be.

**Comment by ethe** [comment]
Wish to support intersection types.

**Comment by HertzDevil** [comment]
> This is possible to implement as ...

The intersection defined this way isn't commutative yet, not even up to equivalent types. Some examples:

```crystal
# covariant Tuple instances
typeof(Intersect({Int32}, {Int32 | Char}).get) # => Tuple(Int32)
typeof(Intersect({Int32 | Char}, {Int32}).get) # => NoReturn

# generic module instances
typeof(Intersect(Indexable(Int32), Enumerable(Int32)).get) # => Indexable(Int32)
typeof(Intersect(Enumerable(Int32), Indexable(Int32)).get) # => NoReturn

# metaclasses between module and including class
typeof(Intersect(File.class, IO.class).get) # => Class
typeof(Intersect(IO.class, File.class).get) # => NoReturn

# metaclasses between superclass and subclass
typeof(Intersect(File.class, IO::FileDescriptor.class).get) # => Class
typeof(Intersect(IO::FileDescriptor.class, File.class).get) # => NoReturn
```

If `Intersect` becomes a built-in type operator it should definitely be commutative in all arguments; the same goes for `Class#&`, which shall be defined in terms of this `typeof` precisely, similar to `#|`. (The union operator is commutative up to equivalent types, but not up to strictly equal types.) For the compiler, this means the following must be true for any `a : Crystal::Type` and `b : Crystal::Type`:

```crystal
c = a.filter_by(b)
d = b.filter_by(a)
if c && d
  c.implements?(d) && d.implements?(c) # C <= D && C >= D
else
  c.nil? && d.nil? # both filtered types are NoReturn
end
```

Doing so would also impact flow typing; for example, this might be possible in the future:

```crystal
x = {1 || 'a'}
if x.is_a?({Int32})
  typeof(x) # => Tuple(Int32)
else
  typeof(x) # => Tuple(Char)
end
```

**Comment by HertzDevil** [comment]
Defining `Class#&` and the `Intersect` operator using the current semantics is the easiest. If we add `&` to AST node restrictions we already face a problem: for what types `A`, `B`, `C` are the following pairs of defs redefinitions?

```crystal
def foo(x : (A & B) | C); end
def foo(x : (A | C) & (B | C)); end

def bar(x : (A | B) & C); end
def bar(x : (A & C) | (B & C)); end
```

It would be good if distributivity holds for any AST node, but I haven't investigated that far yet. Irreducible intersection types are even harder to implement, because if those types are equivalent then it might be preferable to always normalize them to CNFs or DNFs (not doing so would create a situation similar to `NamedTuple` where `T <= U && U <= T` does not imply `T == U`).

**Comment by sol-vin** [comment]
I'm very invested in the idea of a type intersection system. The `module` system in Crystal has a lot of challenges to deal with that Ruby never did. Since Ruby doesn't really do type checking, it's very feasible that if you take some object you could just run a bunch of `.is_a?` methods to find out if it has all the modules required before doing something with it, and even if you didn't, duck typing would let it just work unless it hit something it really didn't have (like a method). Having something in Crystal that allows us to more accurately represent types without creating a bunch if intermediary types would be awesome for creating clear and concise code :)

**Comment by boulme** [comment]
> It would be good if distributivity holds for any AST node, but I haven't investigated that far yet. Irreducible intersection types are even harder to implement, because if those types are equivalent then it might be preferable to always normalize them to CNFs or DNFs (not doing so would create a situation similar to `NamedTuple` where `T <= U && U <= T` does not imply `T == U`).

At some point, you may need to have some BDD-like representation to normalize types. And you will start to embed a SAT-solver within the type-checker.

**Comment by boulme** [comment]
> At some point, you may need to have some BDD-like representation to normalize types. And you will start to embed a SAT-solver within the type-checker.

This remark only applies to an extension of the type system with both intersection and complement. With intersection only, I believe that, as @HertzDevil pointed out, a DNF (or CNF) normalization is sufficient.
